### ISSUE TITLE
search in archives

### ISSUE THREAD

**Issue by flip111** [sharkdp/fd#346]
Would be nice if there was an option to search in archives too (zip, xz, gz, 7z, rar).

**Comment by sharkdp** [comment]
Thank you very much.

I can see how this is a interesting use-case, but I'd rather not add this to `fd`. This would likely add a couple of dependencies and complicate the directory-walking quite a bit.

While not as convenient, note that you can do something like this:
``` bash
fd -e tar.gz -x tar ztvf | grep <pattern>
```
This will search for all `.tar.gz` files, list the contents of each archive (by executing `tar ztvf ...` in parallel on these files) and search for a given pattern

**Comment by flip111** [comment]
Ok i will close this.

However for my curiosity could you explain how it would complicate the directory-walking? And what's against dependencies? If the end-user has a static binary he is happy :smile: 

I'm not sure if with pipe there will be enough leverage like what is possible with native rust. So far i have not seen a search tool that has the ability to search through archives. I think it would be very powerful and also attract more users.

**Comment by sharkdp** [comment]
> Ok i will close this.

**Comment by sharkdp** [comment]
> Ok i will close this.

Oh, we can certainly discuss this ;-)

> However for my curiosity could you explain how it would complicate the directory-walking?

I'm just guessing, but I think we would need to treat archives like folders which also need to be walked (recursively). Right now there is just the file system. Now we have to deal with a whole lot of different archive types and new types of errors (decompression errors), etc.

> And what's against dependencies?

Having more dependencies implies..
* a higher probability of an upstream error
* more work when releasing new versions of `fd` (keeping up to date with updates, fixing things when there are breaking changes, ...)
* more work for package maintainers as `fd` depends on more packages
* larger static executable sizes

**Comment by flip111** [comment]
Maybe it's possible to define some kind of interface / typeclass which can be implemented for an archive type. Maybe somebody has already done this in the rust ecosystem to find the common denominator between archive types (but for searching most archive operations don't need to be supported). Maybe it's also possible to add each archive type as a compile option (like `configure` -- i don't know how rust does this). Then it can be possible to easily enable and disable support for archives. Or just drop support for an archive type if it becomes unmaintainable.

**Comment by tmccombs** [comment]
What is the use-case for this? Are you wanting to search for names in a single archive, or recursively search all archives in the search path as if they were folders?

**Comment by flip111** [comment]
> recursively search all archives in the search path as if they were folders?

this + search normal files

**Comment by Anachron** [comment]
I think the  solution provided in https://github.com/sharkdp/fd/issues/346#issuecomment-431933830 is both neat and fast enough so that we don't need to bundle more dependencies. 

Not sure about other users but I like the "do one thing and do it well" approach from unix a lot.

**Comment by flip111** [comment]
How would that solution work with multiple archive types though? And what
about ergonomics and performance?

**Comment by sharkdp** [comment]
I'm going to close this for now, as this is unlikely to get implemented.

**Comment by berezovskyi** [comment]
In case anyone needs to do such a search in a pinch:

```sh
#!/usr/bin/env bash

# Copyright (c) 2022, Andrii Berezovskyi
# 
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
# SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
# OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

set -uo pipefail


echo "Searching RAW files for '$1'" >&2
fd "$1"

echo "Searching ZIP files" >&2
fd -e zip -x bash -c 'echo -e "\033[1m$2\033[0m" && (unzip -l "$2" 2>&1 | rg "$1")' bash "$1" {} 
echo "Searching 7ZIP files" >&2
fd -e 7z -x bash -c 'echo -e "\033[1m$2\033[0m" && (7z l -ba "$2" 2>&1 | rg "$1")' bash "$1" {} 
echo "Searching TAR.ZST files" >&2
# FAIL, echo goes into grep
# fd -e tar.zst -x echo \; -x tar tvf | rg "$1"
# FAIL, fd is messing with the stderr/stdout
# fd -e tar.zst -x echo >&2 \; -x tar tvf >/dev/null
fd -e tar.zst -x bash -c 'echo -e "\033[1m$2\033[0m" && (tar xvf "$2" 2>&1 | rg "$1")' bash "$1" {} 

echo "Searching TAR.GZ files" >&2
fd -e tar.gz -x bash -c 'echo -e "\033[1m$2\033[0m" && (tar xvf "$2" 2>&1 | rg "$1")' bash "$1" {} 

echo "Search complete" >&2

```

For example, to find all files with the HTML extension:

```
$ fd-ark '.html$'
```

Requires ripgrep, zstd, 7z to be installed – comment out what you don't need.
